libs/mpi/tests/unit/mpi_ring_async_executor.cpp

```diff
@@ -0,0 +1,165 @@
+//  Copyright (c) 2019 John Biddiscombe
+//
+//  SPDX-License-Identifier: BSL-1.0
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#include <hpx/hpx.hpp>
+#include <hpx/hpx_init.hpp>
+#include <hpx/lcos/future.hpp>
+#include <hpx/mpi.hpp>
+#include <hpx/program_options.hpp>
+#include <hpx/runtime/threads/policies/scheduler_base.hpp>
+#include <hpx/runtime/threads/policies/scheduler_mode.hpp>
+#include <hpx/runtime/threads/thread_data.hpp>
+
+#include <array>
+#include <atomic>
+#include <iostream>
+#include <sstream>
+
+#include <mpi.h>
+
+void msg_recv(int rank, int size, int /*to*/, int from, int token, unsigned tag)
+{
+    std::ostringstream temp;
+    temp << "Rank " << std::setfill(' ') << std::setw(3) << rank << " of "
+         << std::setfill(' ') << std::setw(3) << size << " Recv token "
+         << std::setfill(' ') << std::setw(3) << token << " from rank "
+         << std::setfill(' ') << std::setw(3) << from << " tag "
+         << std::setfill(' ') << std::setw(3) << tag;
+    std::cout << temp.str() << std::endl;
+}
+
+void msg_send(int rank, int size, int to, int /*from*/, int token, unsigned tag)
+{
+    std::ostringstream temp;
+    temp << "Rank " << std::setfill(' ') << std::setw(3) << rank << " of "
+         << std::setfill(' ') << std::setw(3) << size << " Sent token "
+         << std::setfill(' ') << std::setw(3) << token << " to   rank "
+         << std::setfill(' ') << std::setw(3) << to << " tag "
+         << std::setfill(' ') << std::setw(3) << tag;
+    std::cout << temp.str() << std::endl;
+}
+
+// this is called on an hpx thread after the runtime starts up
+int hpx_main(hpx::program_options::variables_map& vm)
+{
+    int rank, size;
+    //
+    MPI_Comm_size(MPI_COMM_WORLD, &size);
+    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+
+    if (size < 2)
+    {
+        std::cout << "This test requires N > 1 ranks" << std::endl;
+        return hpx::finalize();
+    }
+
+    //
+    // tell the scheduler that we want to handle mpi in the background
+    // here we use the provided hpx::mpi::poll function but a user
+    // provided function or lambda may be supplied
+    //
+    auto const sched = hpx::threads::get_self_id_data()->get_scheduler_base();
+    sched->set_user_polling_function(&hpx::mpi::poll);
+    sched->add_scheduler_mode(hpx::threads::policies::enable_user_polling);
+    if (rank == 0)
+    {
+        std::cout << "Rank " << std::setfill(' ') << std::setw(3) << rank
+                  << " of " << std::setfill(' ') << std::setw(3) << size
+                  << " scheduler is " << sched->get_description() << "\n\n"
+                  << std::endl;
+    }
+
+    // Ring send/recv around N ranks
+    // Rank 0      : Send then Recv
+    // Rank 1->N-1 : Recv then Send
+
+    hpx::mpi::executor exec(MPI_COMM_WORLD);
+
+    unsigned int const n_loops = 20;
+    std::atomic<int> counter = n_loops;
```

Comment:
If I change the way of using construction, like, `std::atomic<int> counter(n_loops)`, I would not receive compile error. Otherwise, I have a compilaion error:
```
[weile@login3.summit build_Debug]$ make tests.unit.modules.mpi.mpi_ring_async_executor
[  0%] Built target hpx_preprocessor
[  6%] Built target hpx_config
[  6%] Built target hpx_assertion
[  6%] Built target hpx_type_support
[ 12%] Built target hpx_concepts
[ 12%] Built target hpx_filesystem
[ 12%] Built target hpx_format
[ 12%] Built target hpx_timing
[ 18%] Built target hpx_logging
[ 18%] Built target hpx_thread_support
[ 25%] Built target hpx_errors
[ 25%] Built target hpx_basic_execution
[ 31%] Built target hpx_hashing
[ 31%] Built target hpx_concurrency
[ 31%] Built target hpx_datastructures
[ 31%] Built target hpx_debugging
[ 31%] Built target hpx_functional
[ 31%] Built target hpx_naming_base
[ 31%] Built target hpx_serialization
[ 31%] Built target hpx_iterator_support
[ 37%] Built target hpx_util
[ 37%] Built target hpx_topology
[ 37%] Built target hpx_affinity
[ 37%] Built target hpx_allocator_support
[ 37%] Built target hpx_cache
[ 37%] Built target hpx_coroutines
[ 37%] Built target hpx_hardware
[ 37%] Built target hpx_memory
[ 37%] Built target hpx_plugin
[ 43%] Built target hpx_program_options
[ 50%] Built target hpx_threading_base
[ 50%] Built target hpx_synchronization
[ 50%] Built target hpx_threadmanager
[ 50%] Built target hpx_resource_partitioner
[ 50%] Built target hpx_local_lcos
[ 50%] Built target hpx_execution
[ 50%] Built target hpx_statistics
[ 50%] Built target hpx_algorithms
[ 56%] Built target hpx_batch_environments
[ 56%] Built target hpx_checkpoint
[ 62%] Built target hpx_static_reinit
[ 62%] Built target hpx_collectives
[ 62%] Built target hpx_compute
[ 62%] Built target hpx_compute_cuda
[ 62%] Built target hpx_mpi
[ 62%] Built target hpx_resiliency
[ 62%] Built target hpx_segmented_algorithms
[ 62%] Built target hpx_testing
[106%] Built target hpx
[106%] Built target hpx_wrap
[106%] Built target hpx_init
Scanning dependencies of target mpi_ring_async_executor_test
[106%] Building CXX object libs/mpi/tests/unit/CMakeFiles/mpi_ring_async_executor_test.dir/mpi_ring_async_executor.cpp.o
/gpfs/alpine/proj-shared/cph102/weile/dev/src/hpx_mpi_async/hpx/libs/mpi/tests/unit/mpi_ring_async_executor.cpp: In function 'int hpx_main(boost::program_options::variables_map&)':
/gpfs/alpine/proj-shared/cph102/weile/dev/src/hpx_mpi_async/hpx/libs/mpi/tests/unit/mpi_ring_async_executor.cpp:82:32: error: use of deleted function 'std::atomic<int>::atomic(const std::atomic<int>&)'
     std::atomic<int> counter = n_loops;
                                ^~~~~~~
In file included from /gpfs/alpine/proj-shared/cph102/weile/dev/install/boost/1.68.0/include/boost/smart_ptr/detail/spinlock_std_atomic.hpp:19,
                 from /gpfs/alpine/proj-shared/cph102/weile/dev/install/boost/1.68.0/include/boost/smart_ptr/detail/spinlock.hpp:47,
                 from /gpfs/alpine/proj-shared/cph102/weile/dev/src/hpx_mpi_async/hpx/libs/errors/include/hpx/errors/exception_list.hpp:15,
                 from /gpfs/alpine/proj-shared/cph102/weile/dev/src/hpx_mpi_async/hpx/build_Debug/libs/errors/include/hpx/errors.hpp:16,
                 from /gpfs/alpine/proj-shared/cph102/weile/dev/src/hpx_mpi_async/hpx/hpx/runtime/agas/interface.hpp:15,
                 from /gpfs/alpine/proj-shared/cph102/weile/dev/src/hpx_mpi_async/hpx/hpx/runtime/applier/apply.hpp:14,
                 from /gpfs/alpine/proj-shared/cph102/weile/dev/src/hpx_mpi_async/hpx/hpx/apply.hpp:11,
                 from /gpfs/alpine/proj-shared/cph102/weile/dev/src/hpx_mpi_async/hpx/hpx/hpx.hpp:10,
                 from /gpfs/alpine/proj-shared/cph102/weile/dev/src/hpx_mpi_async/hpx/libs/mpi/tests/unit/mpi_ring_async_executor.cpp:7:
/autofs/nccs-svm1_sw/summit/gcc/8.1.1/include/c++/8.1.1/atomic:668:7: note: declared here
       atomic(const atomic&) = delete;
       ^~~~~~
/autofs/nccs-svm1_sw/summit/gcc/8.1.1/include/c++/8.1.1/atomic:672:17: note:   after user-defined conversion: 'constexpr std::atomic<int>::atomic(std::atomic<int>::__integral_type)'
       constexpr atomic(__integral_type __i) noexcept : __base_type(__i) { }
                 ^~~~~~
make[3]: *** [libs/mpi/tests/unit/CMakeFiles/mpi_ring_async_executor_test.dir/mpi_ring_async_executor.cpp.o] Error 1
make[2]: *** [libs/mpi/tests/unit/CMakeFiles/mpi_ring_async_executor_test.dir/all] Error 2
make[1]: *** [libs/mpi/tests/unit/CMakeFiles/tests.unit.modules.mpi.mpi_ring_async_executor.dir/rule] Error 2
make: *** [tests.unit.modules.mpi.mpi_ring_async_executor] Error 2
```

Reply:
@weilewei this should be ok now.